File: scripts/algorithms/N/Next Greater Numerically Balanced Number/Next Greater Numerically Balanced Number.js
/**
 * @param {number} n
 * @return {number}
 */
var nextBeautifulNumber = function(n) {
    //1224444 is next minimum balanced number after 10^6
    for(let i=n+1; i<=1224444;i++){//Sequency check each number from n+1 to 1224444
        if(isNumericallyBalanced(i)){
            return i;//Return the number if is a balanced number
        }
    }
    function isNumericallyBalanced(n){
        let map={},d,nStr=n.toString();
        while(n>0){//Create a map of digits with frequency
            d = n%10;
            if(d>nStr.length){//If we have found a digit greater than the lenght of the number like 227333, here when we see 7, we can return false
                return false;
            }
            if(map[d]===undefined){
                map[d]=1;
            }else{
                map[d]++;
                if(map[d]>d){//If a digit has frequency more than its value, for example 22333344, here when we see fourth 3 than we can return false
                    return false;
                }
            }
            n = Math.floor(n/10);
        }
        for(let key in map){//Check if frequency is equal to the digit
            if(map[key]!==parseInt(key)){
                return false;
            }
        }
        return true;
    }
};

File: scripts/algorithms/N/Next Greater Numerically Balanced Number/Next Greater Numerically Balanced Number.java
// Runtime: 207 ms (Top 46.27%) | Memory: 117.2 MB (Top 26.87%)
class Solution {
    public int nextBeautifulNumber(int n) {

        while(true){
            n++;
           int num = n; //test this number
           int [] freq = new int[10]; // 0 to 9

            while(num > 0){ //calculate freq of each digit in the num
                int rem = num % 10; //this is remainder
                num = num / 10; //this is quotient
                freq[rem] = freq[rem] + 1; //increase its frequency
                if(freq[rem] > rem) break;
            }

            boolean ans = true;

            for(int i = 0;i<10;i++){ //check frequency of each digit
              if(freq[i] != i && freq[i] != 0){
                  ans = false;
                  break;
              }
            }

            if(ans == true){
                return n;
            }
        }
    }
}